src/hypothesis/internal/cathetus.py

```diff
@@ -0,0 +1,58 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2018 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from math import sqrt, isnan, isinf, fabs
+from sys import float_info as fi
+
+
+def cathetus(h, a):
+    """Given the lengths of the hypotenuse and a side of a right
+    triangle, return the length of the other side.  A companion
+    to the C99 hypot() function.  Some care is needed to avoid
+    underflow in the case of small arguments, and overflow in the
+    case of large arguments as would occur for the naive implementaion
+    as sqrt(h*h - a*a).  The behaviour with respect the non-finite
+    arguments (NaNs and infinities) is designed to be as consistent
+    as possible with the C99 hypot() specifications.
+
+    Based on the C99 implementation https://github.com/jjgreen/cathetus
+    """
+    if isnan(h):
+        return float(u'nan')
+
+    if isinf(h):
+        if isinf(a):
```

Comment:
What should happen if `isnan(a)`?  I suspect we should return `nan` rather than `inf`.

Reply:
At first blush, I would have thought so too "everything than NaN touches becomes a NaN", but C99 mandates of `hypot` "if one of the arguments is ±∞, hypot returns +∞ even if the other argument is NaN", which kind-of make sense if we think of NaN as "any or all numbers".  And 

    cathetus(inf, inf) = sqrt(inf^2 - inf^2) = NaN, 

naturally, so this behaviour makes `cathetus(hypot(a, b), b) == a` when `a = NaN` and `b = inf`.

I have a feeling that this "unexpected consistency" is due to the C99 committee originally planning to have `cathetus` in the standard, but dropped it though lack of perceived utility.

Reply:
Ah, that makes sense!  Maybe add a short comment to this effect?

Reply:
done